testing: (onnx) onnx-mlir MNIST model targeting

kayode-gif opened the pull request:
resolves #2243

superlopuh commented on tests/filecheck/projects/onnx/mnist.mlir:
```diff
@@ -0,0 +1,45 @@
+module attributes {llvm.data_layout = "e-m:o-i64:64-i128:128-n32:64-S128", llvm.target_triple = "arm64-apple-darwin23.1.0", "onnx-mlir.symbol-postfix" = "mnist"} {
+  func.func @main_graph(%arg0: tensor<1x1x28x28xf32> loc(unknown)) -> tensor<1x10xf32> attributes {input_names = ["input.1"], output_names = ["19"]} {
+    %0 = onnx.Constant dense<[-1, 320]> : tensor<2xi64> loc(#loc1)
+    %1 = onnx.Constant dense<"0x9FD2AE3D73CA0DBE3A7EF2BC04A243BEA47F59BEBE0C8E3E219FC8BEFC950CBEFA094FBEA36F9F3D3E88933E9ECB953ECE9EF23EA986293F1286983EB077B93CDE2A563E654E6F3E3399833CCF84F63D67162DBED24424BEBE2611BFB271D6BE4688C8BDE9CEC03B9A8A613EF46AA23DE01F823EECD82B3CE93F333E5F9BF4BCA8BF133EE4F9AABD30E0A13C36CD3BBEF07A5DBEFC8CA1BE28D99EBE2794CEBD4604DCBD4F6F1C3E1DC0A0BE759A453CA8E2B83B307D9AB ...
```
```suggestion
    %1 = onnx.Constant dense<1.0> : tensor<10x1x5x5xf32> loc(#loc2)
```

superlopuh commented on tests/filecheck/projects/onnx/mnist.mlir:
```diff
@@ -0,0 +1,45 @@
+    %0 = onnx.Constant dense<[-1, 320]> : tensor<2xi64> loc(#loc1)
+    %1 = onnx.Constant dense<"0x9FD2AE3D73CA0DBE3A7EF2BC04A243BEA47F59BEBE0C8E3E219FC8BEFC950CBEFA094FBEA36F9F3D3E88933E9ECB953ECE9EF23EA986293F1286983EB077B93CDE2A563E654E6F3E3399833CCF84F63D67162DBED24424BEBE2611BFB271D6BE4688C8BDE9CEC03B9A8A613EF46AA23DE01F823EECD82B3CE93F333E5F9BF4BCA8BF133EE4F9AABD30E0A13C36CD3BBEF07A5DBEFC8CA1BE28D99EBE2794CEBD4604DCBD4F6F1C3E1DC0A0BE759A453CA8E2B83B307D9AB ...
+    %2 = onnx.Constant dense<[0.143770665, 0.0425382107, -0.0444311202, 0.0654672608, -0.0245875381, -0.00526585756, 0.103917263, -0.379304767, 0.0496097282, 0.166007519]> : tensor<10xf32> loc(#loc3)
+    %3 = onnx.Constant dense<"0x798AF53D7D3AC43D7B9FD03CCE97763C52D932BD3E24DD3D19AB9D3D586C9A3D3507C23DA81EC23D65094ABD2B20993D9A8DE23D42AA593EB1020C3E02A5C1BDE88DF7BD8DFE81BB8D54153EEDA7C13D294DBABDBDED553C3663E53D725C13BDDB175A3B75FC91BCA427893D2DB3B23DB956693CB5AD3B3DDEFAC7BCCBAC943D9091B73D1C28403DAD26CE3CB486113CCD899ABD02CBC23CE95DB43AD510CF3D0A5D59BD717DFCBCC973883DD5E71B3D8EE6573D6E8C553 ...
```
```suggestion
    %3 = onnx.Constant dense<1.0> : tensor<20x10x5x5xf32> loc(#loc4)
```

superlopuh commented on tests/filecheck/projects/onnx/mnist.mlir:
can you please add filecheck lines? A RUN and the test output that we currently get?

kayode-gif replied:
it seems it doesn't like the `loc` bits at times, i assume thats okay to remove? and also i may have to reformat to our version of ```onnx.constant``` since it doesnt like the 
`onnx.constant dense<>` format

superlopuh replied:
Yeah our loc support isn't great, let's drop them for now

superlopuh replied:
let's fix our format

superlopuh replied:
Ah I see they hack the lock and use it as the node name...

superlopuh replied:
This is potentially a problem. I'd say let's do our best to make it work as-is

superlopuh replied:
To clarify: I'd say let's use each line where our parser fails to file an issue with the underlying error, and solve it before moving on to the lowering from onnx to linalg. If the loc handling is a blocker, let's take a look at that

tobiasgrosser replied:
Do we really want to add such a large file to the repo?

superlopuh replied:
We don't care about the weight values, so we can reduce them all to `dense<1.0>`, it shouldn't be large after those changes

superlopuh commented on tests/filecheck/projects/onnx/mnist.mlir:
```diff
@@ -0,0 +1,45 @@
+    %2 = onnx.Constant dense<[0.143770665, 0.0425382107, -0.0444311202, 0.0654672608, -0.0245875381, -0.00526585756, 0.103917263, -0.379304767, 0.0496097282, 0.166007519]> : tensor<10xf32> loc(#loc3)
+    %3 = onnx.Constant dense<1.0> : tensor<20x10x5x5xf32> loc(#loc4)
+    %4 = onnx.Constant dense<[0.0641330257, -0.0504570939, 0.0234062821, 0.0211601146, 0.066841349, 0.0223691612, 0.0544225238, 0.00133739749, -0.0348719358, 0.0357048325, -0.0534538552, -0.00560635189, -0.0557079688, -0.0390234925, -0.0576108806, 0.0383031704, -0.064689137, 0.0698348656, -0.010613976, -0.0297186933]> : tensor<20xf32> loc(#loc5)
+    %5 = onnx.Constant dense<"0xEF19333DD820E9BC044F193DC2049C3D2E4F1B3DC2142EBDA360B0BD8DCBCEBD6B6BB7BDC8CA38BD3E95B1BD503D27BC35D5ADBC6398A6BC56AD233DFA6B533DC9DA63BDB78FB7BB2D29D93CB43E50BD0A12F23C2DD621BD266873BD5BD36BBDB8D204BB9287003E5B367FBDF8A3E33CE4129D3BCD031C3D03619A3CCEB9763D991DDA3B45EFF7BC6020233D17C393BDBB3D003DF629633D6FADD9BB637D8CBDC283D6BA74869FBCB97FF43CFC9D273D6D1AB2BBEB4BEDB ...
```
Can you squash this and below to 1.0 also, please?

compor reviewed (approved):
The `loc` values seem to be used as an organizational marker to "track" the group of operations; we can just use a comment.

Simplify the huge constants and it's good for me to use as test case for whatever we are missing in our parsing support.

superlopuh commented:
Ah yes indeed, maybe we can just throw them away

kayode-gif commented:
I have made small changes within this PR:
1. added dilatrions attribute to `onnx.maxpoolsingleout`, in ours its required, so just adding the default value makes no difference, since that is what the model omits
2. `onnx.Gemm `was throwing incompatible shapes error, seems the transpose code wasn't doing what i thought, fixed it
3. changed the attribute value in the lowering to match with our updated constant format

superlopuh commented on xdsl/transforms/convert_onnx_to_linalg.py:
```diff
@@ -93,7 +93,7 @@ def make_unique_name(self):
     @op_type_rewrite_pattern
     def match_and_rewrite(self, constant: onnx.Constant, rewriter: PatternRewriter, /):
-        attr_value = list(constant.attributes.values())[1]
+        attr_value = constant.value
```
get_value()?

kayode-gif replied:
> get_value()?
helper function? you mean

kayode-gif replied:
sorry, i am confused what you mean

superlopuh replied:
Yep, whatever the type of the value, we should create a global with it. This code only gets the default `dense` of floats, but if it's ints it'll break

kayode-gif replied:
of so parse the attr_value type into the global?

superlopuh commented:
Closing this for now, we can reopen when there's more work on the ONNX side of things
